draft: `commandStart` more

adomani opened the pull request:
---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

adomani committed:
some changes

adomani committed:
chore: a bunch of whitespace fixes

adomani committed:
some modifications to the linter

adomani committed:
chore: whitespace after ∀

adomani committed:
chore: whitespace after ∃

adomani committed:
Merge remote-tracking branch 'upstream/master' into commandStart_more

adomani committed:
chore: fix whitespaces

adomani committed:
more test

adomani committed:
two more exceptions, fix most tests

adomani committed:
ignore `throwError`

adomani committed:
chore: more whitespace fixes

adomani committed:
chore: space following `Σ`

adomani committed:
chore: space following `Σ`

adomani committed:
Merge branch 'master' of github.com:leanprover-community/mathlib4 into commandStart_more

adomani committed:
also `Σ'`

adomani committed:
more sigmas

adomani committed:
another batch

adomani committed:
silence macro and elab

adomani committed:
add space to `fastinstance%`

adomani committed:
omit term-mode have

adomani committed:
long line

adomani committed:
Merge branch 'whitespace/space_after_sigma' into commandStart_more

adomani committed:
chore: another batch of space

adomani committed:
Merge branch 'master' of github.com:leanprover-community/mathlib4 into commandStart_more

adomani committed:
further silences

adomani committed:
Merge branch 'master' of github.com:leanprover-community/mathlib4 into commandStart_more

adomani committed:
chore: `n+1` to `n + 1`

adomani committed:
chore: `n+1` to `n + 1`

adomani committed:
Merge branch 'master' of github.com:leanprover-community/mathlib4 into commandStart_more

adomani committed:
long lines

adomani committed:
fix

adomani committed:
fixes

adomani committed:
more fixes

adomani committed:
two more fixes

adomani committed:
1-z

adomani committed:
Merge branch 'whitespace/nplusone_to_n_plus_one' into commandStart_more

adomani committed:
more fixes

adomani committed:
fixes

adomani committed:
again?

adomani committed:
another

adomani committed:
reverse

adomani committed:
reverse

adomani committed:
fix

adomani committed:
FLT

adomani committed:
ppSpace in tensor product

adomani committed:
chore: another collection of whitespace fixes

adomani committed:
conditional

adomani committed:
gcongr ppSpace

adomani committed:
ignore condExp

adomani committed:
chore: another set of whitespace fixes

adomani committed:
fixes

adomani committed:
induction with

adomani committed:
namespace, simp! example

adomani committed:
more fixes

adomani committed:
one isolated

adomani committed:
Merge branch 'master' of github.com:leanprover-community/mathlib4 into commandStart_more

adomani committed:
matrix invertible

adomani committed:
powerset ppSpace

adomani committed:
chore: another batch of fixes

adomani committed:
chore: spaces fixes
Gigantic batch

adomani committed:
omit set-builder

adomani committed:
Merge branch 'master' of github.com:leanprover-community/mathlib4 into commandStart_more

adomani committed:
ppSpace Finset.smulFinset ppSpace

adomani committed:
chore: small fixes

adomani committed:
Add space after `∀'` and `∃'`

adomani committed:
superfactorial notation

adomani committed:
set_option

adomani committed:
chore: a small batch of spaces

adomani committed:
chore: a small batch of spaces
Including some Lie algebras

adomani committed:
Merge branch 'master' of github.com:leanprover-community/mathlib4 into commandStart_more

adomani committed:
inv, =>, *|*

adomani committed:
` )`

adomani committed:
Odd syntax change for upperscripts only on R, not Spheres

adomani committed:
chore: whitespace changes
Found by #26926.

Fixing from the bottom.

adomani commented on Mathlib/NumberTheory/JacobiSum/Basic.lean:
```diff
@@ -200,7 +200,7 @@ lemma jacobiSum_mul_jacobiSum_inv (h : ringChar F' ≠ ringChar F) {χ φ : MulC
     (hφ : φ ≠ 1) (hχφ : χ * φ ≠ 1) :
     jacobiSum χ φ * jacobiSum χ⁻¹ φ⁻¹ = Fintype.card F := by
   obtain ⟨n, hp, hc⟩ := FiniteField.card F (ringChar F)
-  let ψ := FiniteField.primitiveChar F F' h   -- obtain primitive additive character `ψ : F → FF'`
```
I went by [#mathlib4 > Linting space around variables @ 💬](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/Linting.20space.20around.20variables/near/504697624): maybe you can raise the question again on Zulip?

Special casing this kind of use will be exceedingly hard for this linter, though.

MichaelStollBayreuth replied:
I'm not saying it should not be changed. It is probably better to not special case things. But I noticed that the extra spaces were not removed in the following line, so this looks inconsistent to me.

adomani replied:
Ah, the reason for this is that there is an inconsistency with how `let X ...` and `let (X)` pretty-print: the former has a space between `let` and `X`, the latter does *not* have a space between `let` and `(X)`.  When I noticed this, I removed `let` from what the linter inspects, but I had apparently already changed the first of these lines.  I'll remove the following space as well, since the plan is to "fix" `let`, not leave it special-cased forever!

Thanks!

adomani commented:
> I'm not sure it's a good idea to insist on spaces around `+` in docstrings (and then why not also require spaces around `*`, `^`, etc.?).

I am planning to change the others as well: see the discussion at #26937.

adomani committed:
Michaels suggestion

adomani committed:
Jovan's catch

adomani committed:
fix Akra-Bazzi

adomani committed:
more with more Akra-Bazzi

adomani committed:
remove calc parens

adomani committed:
remove verbose option

adomani committed:
heavy files

adomani committed:
notation, no change in pretty-printing

adomani committed:
notation for pretty-printing

adomani committed:
more large numbers

adomani committed:
a small batch

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

adomani committed:
Merge branch 'master' of github.com:leanprover-community/mathlib4 into commandStart_more

adomani committed:
revert syntax change

adomani committed:
revert syntax change

adomani committed:
unlint syntax

adomani committed:
unlint syntax

adomani committed:
Merge branch 'whitespace/with_inv' into commandStart_more

adomani committed:
Merge branch 'master' of github.com:leanprover-community/mathlib4 into whitespace/nplusone_to_n_plus_one

adomani committed:
revert doc-string changes

adomani committed:
final round of corrections: thanks Michael!

adomani committed:
Merge branch 'whitespace/nplusone_to_n_plus_one' into commandStart_more

adomani committed:
Merge branch 'master' of github.com:leanprover-community/mathlib4 into commandStart_more

adomani committed:
Merge branch 'master' of github.com:leanprover-community/mathlib4 into whitespace/with_inv

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

adomani committed:
Merge branch 'master' of github.com:leanprover-community/mathlib4 into commandStart_more

adomani committed:
parallelScanAux regenerates the "correct" string.

adomani committed:
docs

adomani committed:
chore: remove space after ⅟

adomani committed:
chore: remove space after ⅟

adomani committed:
two more

adomani committed:
Merge branch 'master' of github.com:leanprover-community/mathlib4 into whitespace/with_inv

adomani committed:
Merge branch 'whitespace/with_inv' into commandStart_more

adomani committed:
refix

adomani committed:
reversals

adomani committed:
Merge branch 'whitespace/with_inv' into commandStart_more

adomani committed:
Merge branch 'master' of github.com:leanprover-community/mathlib4 into whitespace/with_inv

adomani committed:
Merge branch 'whitespace/with_inv' into commandStart_more

adomani committed:
Merge branch 'master' of github.com:leanprover-community/mathlib4 into commandStart_more

adomani committed:
leadTrail and splaying, docs

adomani committed:
more syntax comparisong, add rebuild use substring

adomani committed:
Merge branch 'master' of github.com:leanprover-community/mathlib4 into commandStart_more

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

adomani committed:
Merge branch 'master' of github.com:leanprover-community/mathlib4 into commandStart_more

adomani committed:
more attempts

adomani committed:
add consumeIgnoring

adomani committed:
working? some docs

adomani committed:
Merge branch 'master' of github.com:leanprover-community/mathlib4 into commandStart_more

adomani committed:
docs

adomani committed:
more docs and renames

adomani committed:
getting there

adomani committed:
better parsing for `#mex`, add `getExceptions`

adomani committed:
minor adjustments

adomani committed:
make pretty more robust, forceSpaceAfter, accumulate kinds

adomani committed:
allow black out syntaxnodekinds

adomani committed:
move around, more exceptions

adomani committed:
Add `ExcludedSyntaxNodeKind`, improve `#mex`, mkWindowSubstring

adomani committed:
docs

adomani committed:
#mex reports reported and unreported

adomani committed:
move #mex, add `mkWindowSubstring'`

adomani committed:
more exclusions

adomani committed:
fix some tests

adomani committed:
more exceptions, processing bugs, `#mex`

adomani committed:
further exclusions

adomani committed:
more test fixes

adomani committed:
arrays and lists

adomani committed:
where and anonymous constructor

adomani committed:
more test fixes

adomani committed:
syntax

adomani committed:
exit in test

adomani committed:
ignoreSpaceAfter, partition of mex, silence linter file

adomani committed:
more guard msgs in tests

adomani committed:
aesop ignore

adomani committed:
more tests

adomani committed:
notation3, library_note, parse guillemets

adomani committed:
more aesop exceptions, better verbose setting

adomani committed:
expected window and tests

adomani committed:
fix data.list.basic

adomani committed:
deal more with choice

adomani committed:
adapt tests

adomani committed:
more exclusions, toLinterWarning also in `#mex`

adomani committed:
priority in `AList`

adomani committed:
unset option in `#mex`

adomani committed:
add `getChoiceNode`, reset state after `insertSpaces`

adomani committed:
escape sed

adomani committed:
typo

adomani committed:
another typo

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

adomani committed:
break, mkStrings, a few more structures

adomani committed:
function

adomani committed:
mkRangeError

adomani committed:
mkWdw

adomani committed:
add (some) syntaxnode awareness

adomani committed:
fix whitespace after `where`

adomani committed:
reformat in test and some small adjustments

adomani committed:
Merge branch 'master' of github.com:leanprover-community/mathlib4 into commandStart_more

adomani committed:
docs

adomani committed:
remove import

grunweg committed:
Merge branch 'master' into commandStart_more

grunweg committed:
chore: fix spaces in rintro and rcases patterns

use _ | _ instead of _|_.
I'm not yet sure if I prefer "(_ | _)" over "(_|_)"; to be discussed!"

grunweg committed:
Spacing around not: clearly good

grunweg committed:
Less sure about those

grunweg committed:
Some misc good changes

grunweg commented:
I'm a bit on the fence about some changes in [6d04d61](https://github.com/leanprover-community/mathlib4/pull/26926/commits/6d04d6175fade6f7b8ec2ef1093eda0529f78a6f), and actively unsure about [8bf1b2d](https://github.com/leanprover-community/mathlib4/pull/26926/commits/8bf1b2d2f32fb9320da7b71dcb216f8fb46fcd17). (My ideal spacing around \not is more nuanced than "never a space after".)

grunweg commented:
Some bug reports as I looked over some of the remaining errors
- a bug in the "choose the right output window" logic, as exemplified by the line
`example : ¬ UnivLE.{u+1, u} := sorry` in Logic/UnivLE:
the desired output is `example : ¬UnivLE.{u + 1, u} := sorry` (remove one space, add two),
but the linter warnings are these --- the errors are good, but the suggestions are very confusing!

Warning: warning: Mathlib/Logic/UnivLE.lean:80:11: remove space in the source
This part of the code
  '¬ UnivLE.{u+1,'
should be written as
  '¬UnivLE.{u'
Note: This linter can be disabled with `set_option linter.style.commandStart false`
Warning: warning: Mathlib/Logic/UnivLE.lean:80:21: add space in the source
This part of the code
  'UnivLE.{u+1,'
should be written as
  '¬UnivLE.{u +'
Note: This linter can be disabled with `set_option linter.style.commandStart false`
Warning: warning: Mathlib/Logic/UnivLE.lean:80:22: add space in the source
This part of the code
  'UnivLE.{u+1,'
should be written as
  '+ 1,'
Note: This linter can be disabled with `set_option linter.style.commandStart false`


- no-op suggestions
Warning: warning: Mathlib/Data/List/TFAE.lean:60:9: add space in the source
This part of the code
  'have :='
should be written as
  'have :='
Note: This linter can be disabled with `set_option linter.style.commandStart false`
The suggestions are identical (and the current and desired state look fine to me).

- lots of warnings of the kind `warning: no change at [line]`, with output including
```
Maybe because the `SyntaxNodeKind`s contain:
hygieneInfo: true
choice: false
```
